x/archivedb/key.go

```diff
@@ -12,77 +13,135 @@ import (
+type databaseKey interface {
+	InnerKey() []byte
+	Bytes() []byte
+	Height() uint64
+}
+
+// dbKey
 //
 // The keys contains a few extra information alongside the given key. The key is
 // what is known outside of this internal scope as the key, but this struct has
-// more information compacted inside the keyInternal to take advantage of natural
+// more information compacted inside the dbKey to take advantage of natural
 // sorting. The inversed height is stored right after the Prefix, which is
-// MaxUint64 minus the desired Height. An extra byte is also packed inside the
-// key to indicate if the key is a deletion of any previous value (because
-// archivedb is an append only database).
+// MaxUint64 minus the desired Height.
 //
-// Any other property are to not serialized but they are useful when parsing a
-// keyInternal struct from the database
-type keyInternal struct {
-	key       []byte
-	height    uint64
-	isDeleted bool
+// All keys are prefixed with a VarUint64, which is the length of the keys, the
+// idea is to make reads even simpler, making it effectively a O(1) operation.
+// Before this change it would have been O(1) unless a another key existed with
+// the same prefix as the requested. By appending the length of the inner key as
+// the first bytes, it is not longer possible to query by a prefix (the full key
+// is needed to do a lookup) but all reads are a O(1).
+type dbKey struct {
+	key    []byte
+	height uint64
 }
 
 // Creates a new Key struct with a given key and its height
-func newInternalKey(key []byte, height uint64) *keyInternal {
-	return &keyInternal{
-		key:       key,
-		isDeleted: false,
-		height:    height,
+func newKey(key []byte, height uint64) *dbKey {
+	return &dbKey{
+		key:    key,
+		height: height,
 	}
 }
 
-func (k *keyInternal) Bytes() []byte {
+func (k *dbKey) Bytes() []byte {
 	keyLen := len(k.key)
-	bytes := make([]byte, keyLen+internalKeySuffixLen+internalKeyPrefixLen)
-	copy(bytes[0:], internalKeyPrefix)
-	copy(bytes[internalKeyPrefixLen:], k.key)
-	binary.BigEndian.PutUint64(bytes[internalKeyPrefixLen+keyLen:], math.MaxUint64-k.height)
-	if k.isDeleted {
-		bytes[internalKeyPrefixLen+keyLen+longLen] = 1
-	} else {
-		bytes[internalKeyPrefixLen+keyLen+longLen] = 0
-	}
+	keyLenBytes := binary.AppendUvarint([]byte{}, uint64(keyLen))
+
+	bytes := make([]byte, len(keyLenBytes)+keyLen+internalKeySuffixLen)
+	offset := copy(bytes[0:], keyLenBytes)
+	offset += copy(bytes[offset:], k.key)
+	binary.BigEndian.PutUint64(bytes[offset:], math.MaxUint64-k.height)
 	return bytes
 }
 
-// Takes a slice of bytes and returns a Key struct
-func parseKey(rawKey []byte) (*keyInternal, error) {
-	var key keyInternal
-	if internalKeySuffixLen+internalKeyPrefixLen >= len(rawKey) {
+func (k *dbKey) Height() uint64 {
+	return k.height
+}
+
+func (k *dbKey) InnerKey() []byte {
+	return k.key
+}
+
+type dbMetaKey struct {
+	key []byte
+}
+
+func newMetaKey(key []byte) *dbMetaKey {
+	return &dbMetaKey{key}
+}
```

Comment:
Does it really make sense for this to be a whole type? I feel like `newMetaKey` could just return `[]byte` which is the current output of `.Bytes()`

Reply:
I think we should be able to guarantee it will never be picked up by the iterator